### PR TITLE
Mapgen: Run light spread algorithm iteratively

### LINKED ISSUE #4399: Mintest crashes when vm:set_lighting({day = 7, night = 7}) or higher
I found a strange bug. I modified the [intersect light](https://forum.minetest.net/viewtopic.php?f=9&t=13414) mod from duane a little bit and set vm:set_lighting({day = 15, night = 15}) to see what happens. Everything works fine, also if I set vm:set_lighting({day = 0, night = 0})

But if I set vm:set_lighting({day = 7, night = 7}) minetest crashes after a few seconds with a memory error. It also crashes with values of 8, 9, 10 and 11, but it takes longer the higher the value is. Maybe it will crash with 15 too, but I never played that long

I had no crash below 7 or it takes extremely long and I didn't try any number yet

Edit: I use the latest dev version from Ubuntu PPA, running on Ubuntu 16.04 64 bit

### LINKED ISSUE #7864: There's something in the attached map that causes a crash with no logged cause
##### Issue type
 - Bug report
 
##### Minetest version

```
minetest-5.0.0-0e306c0-win64 (sfan5's build)
```

##### OS / Hardware

Operating system: Windows 10 Home
CPU: Intel i7-3770K @ 3.5GHz, 8.00 GB ram

##### Summary

I've been working on a mod that, among other things, adds volcanic mountains to any existing mapgen. I have managed to use this mod to generate a particular map block in it that causes Minetest to hard-crash as soon as it attempts to emerge the map block, with no useful information in the debug log (even at verbose) about exactly what it was that caused the crash. I've attached a world folder to this issue with a known example of one of these "poison" map blocks, but this bug can be reproduced in earlier versions of Minetest (I've seen this crash happen in my old 0.4.16 install) and is very consistent whenever I find an example.

Once the world has been generated, my mapgen mod is no longer required - it can be disabled and the crash still happens. No special blocks are being added to the map, nor any fancy metadata.

##### Steps to reproduce

To reproduce the crash in the existing attached world [volc.zip](https://github.com/minetest/minetest/files/2579232/volc.zip):

Join the world in singleplayer mode as "singleplayer". Singleplayer should be located at -1949, 79, 657, adjacent to a large stone mountain with gravel patches on its flanks (this is the volcano generated by my mod). Look in the direction of the mountain and within a few seconds the game will crash. If the logging is in verbose mode the last line of the log will read:

```
INFO[Emerge-0]: initialEmerge: area: (-2048,32,752)(-1937,143,863)=112x112x112=1404928 (5MB)
```

This has been consistent across multiple reproduction attempts, when the game tries to emerge that area it crashes.

The specific version of my mod that was used to generate this map was:

https://github.com/FaceDeer/magma_conduits/tree/e749dfe39f2bb7a871fd8de17bd1d2faa34188a2

With the following settings in minetest.conf to disable various irrelevant features:

```
magma_conduits_magma_veins = false
magma_conduits_glowing_rock = false
magma_conduits_cook_soil = false
magma_conduits_remove_default_lava = false
```

But it should be possible to get this crash to happen without the magma_conduits mod actually being active, since its work is apparently done at this point.

If you want to reproduce this entirely from scratch, create a new world with the name "volc", seed "1", mapgen v7, and Minetest Game. Enable the magma_conduits mod, start playing, /grant singleplayer all, and then /teleport -1949, 79, 657. Look toward the cursed mountain and perish.

Edit: Another volcano that produces a crash in this world is located at 1540, 0, 2760.

### LINKED ISSUE #8763: Segmentation fault in all-lava game.
##### Minetest version

```
Minetest 5.1.0-dev-debug-e788ee28 (Linux)
Using Irrlicht 1.8.4
```

##### Summary
The game segfaults after a few moments of activity in the all-lava game. @jastevenson303 has the same issue.

##### Steps to reproduce
Get the lavaland game by @jastevenson303 https://github.com/jastevenson303/lavaland
I used commit `b0d79cc9b7784b096b685b997a9705a4148edefa`.

Create a new world with that game.
Load it up and splash around, respawn, etc. for a bit. After some time (note the slow mapgen) the game will segfault. This occurs in both debug and release builds.

````
Thread 25 "Emerge-0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff9c88f700 (LWP 23018)]
0x0000555555c18993 in VoxelArea::index (this=0x7fff848d8018, p=...) at /home/USER/src/minetest/src/voxel.h:277
277			return index(p.X, p.Y, p.Z);
````

Notably, removing the `paramtype` and `light_source` fields from the lava nodes' definitions will prevent the crash.
Also notably, setting `ulimit -s` from `8192` to `262144` prevents the crash for me.

My guess is a stack overflow due to sheer number of light calculations.

~~@jastevenson303 has said this does not occur when hosting a game, only in singleplayer.~~

##### Backtrace
This is the first page of the backtrace, it continues with Mapgen::lightSpread as far as I can go.
````
#0  0x0000555555c18993 in VoxelArea::index (this=0x7fff848d8018, p=...) at /home/USER/src/minetest/src/voxel.h:277
No locals.
#1  0x0000555555daa324 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=222 '\336') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:443
        vi = 2617835648
        n = @0x7fff9c090080: {param0 = 240, param1 = 9 '\t', param2 = 156 '\234'}
        light_day = 0 '\000'
        light_night = 0 '\000'
#2  0x0000555555daa462 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=222 '\336') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:470
        vi = 196263
        n = @0x7fff8674705c: {param0 = 32, param1 = 222 '\336', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 0 '\000'
#3  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=31 '\037') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 208807
        n = @0x7fff8675345c: {param0 = 32, param1 = 31 '\037', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 16 '\020'
#4  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=47 '/') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 221351
        n = @0x7fff8675f85c: {param0 = 32, param1 = 47 '/', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 32 ' '
#5  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=63 '?') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 233895
        n = @0x7fff8676bc5c: {param0 = 32, param1 = 63 '?', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 48 '0'
#6  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=79 'O') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 246439
        n = @0x7fff8677805c: {param0 = 32, param1 = 79 'O', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 64 '@'
#7  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=95 '_') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 258983
        n = @0x7fff8678445c: {param0 = 32, param1 = 95 '_', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 80 'P'
#8  0x0000555555daa546 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=111 'o') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:473
        vi = 271527
        n = @0x7fff8679085c: {param0 = 32, param1 = 111 'o', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 96 '`'
#9  0x0000555555daa592 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=127 '\177') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:474
        vi = 271639
        n = @0x7fff86790a1c: {param0 = 32, param1 = 127 '\177', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 112 'p'
#10 0x0000555555daa462 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=143 '\217') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:470
        vi = 259095
        n = @0x7fff8678461c: {param0 = 32, param1 = 143 '\217', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 128 '\200'
#11 0x0000555555daa462 in Mapgen::lightSpread (this=0x7fff9c88dfc0, a=..., p=..., light=159 '\237') at /home/USER/src/minetest/src/mapgen/mapgen.cpp:470
        vi = 246551
        n = @0x7fff8677821c: {param0 = 32, param1 = 159 '\237', param2 = 0 '\000'}
        light_day = 14 '\016'
        light_night = 144 '\220'
````

### PR DESCRIPTION
**Pull request by sfan5**
fixes #4399, fixes #7864, fixes #8763

no more segfaults and in the practically impossible event that all your memory runs out you get an `std::bad_alloc` instead

@paramat does the order in which light spreading runs matter? (it doesn't appear to from my testing and from how I understood the algorithm)
Additional lightSpread() "calls" will go to the end of the queue, which means older "calls" run before.

## To do

This PR is Ready for Review.

## How to test

Use lavaland subgame as described in issue 8763

### CODE REVIEW

File: src/mapgen/mapgen.cpp [hunk 1]
```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
...
+		// we hit a solid block that light cannot pass through.
+		if ((light_day  <= (n.param1 & 0x0F) &&
+			light_night <= (n.param1 & 0xF0)) ||
+			!ndef->get(n).light_propagates)
+			return;
+
+		// Since this recursive function only terminates when there is no light from
+		// either bank left, we need to take the max of both banks into account for
+		// the case where spreading has stopped for one light bank but not the other.
+		light = MYMAX(light_day, n.param1 & 0x0F) |
```
Comment (ClobberXD) [inline src/mapgen/mapgen.cpp]: Why not assign the rvalue to `n.param1` directly?

Comment (sfan5) [reply src/mapgen/mapgen.cpp]: code
doesn't look much better, does it?

Comment (ClobberXD) [reply src/mapgen/mapgen.cpp]: True, nvm :)

**Comment by DS-Minetest** [comment]
Wouldn't it be better to fill a queue in `Mapgen::spreadLight` (not `Mapgen::lightSpread`) with the high light values and then empty this one queue? This way, things like lava lakes would be calculated much faster.

**Comment by HybridDog** [comment]
I think that there are algorithms which could work faster.
https://github.com/minetest/minetest/issues/8763#issuecomment-519551783
http://www.adammil.net/blog/v126_A_More_Efficient_Flood_Fill.html
https://en.wikipedia.org/wiki/Flood_fill#Scanline_fill
The last two may not work in 3D.

**Comment by sfan5** [comment]
If someone wants to implement a better algorithm they can open another PR, it's out of scope here.

File: src/mapgen/mapgen.cpp [hunk 2]
```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
...
+		u8 light_night = light & 0xF0;
+		if (light_night > 0)
+			light_night -= 0x10;
 
-	n.param1 = light;
+		// Bail out only if we have no more light from either bank to propogate, or
+		// we hit a solid block that light cannot pass through.
+		if ((light_day  <= (n.param1 & 0x0F) &&
+			light_night <= (n.param1 & 0xF0)) ||
+			!ndef->get(n).light_propagates)
```
Comment (DS-Minetest) [inline src/mapgen/mapgen.cpp]: Please indent twice here:
```cpp
		if ((light_day  <= (n.param1 & 0x0F) &&
				light_night <= (n.param1 & 0xF0)) ||
				!ndef->get(n).light_propagates)
			continue;
```

File: src/mapgen/mapgen.cpp [hunk 3]
```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
...
+				!ndef->get(n).light_propagates)
+			continue;
 
-	lightSpread(a, p + v3s16(0, 0, 1), light);
-	lightSpread(a, p + v3s16(0, 1, 0), light);
-	lightSpread(a, p + v3s16(1, 0, 0), light);
-	lightSpread(a, p - v3s16(0, 0, 1), light);
-	lightSpread(a, p - v3s16(0, 1, 0), light);
-	lightSpread(a, p - v3s16(1, 0, 0), light);
+		// Since this recursive function only terminates when there is no light from
```
Comment (SmallJoker) [inline src/mapgen/mapgen.cpp]: It's no longer a function.

File: src/mapgen/mapgen.cpp [hunk 4]
```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
...
+			p = item.first;
+			light = item.second;
+		}
+		queue.pop_front();
 
-	// Decay light in each of the banks separately
-	u8 light_day = light & 0x0F;
-	if (light_day > 0)
-		light_day -= 0x01;
+		if (light <= 1 || !a.contains(p))
```
Comment (SmallJoker) [inline src/mapgen/mapgen.cpp]: `n.param1` also needs to be updated for `light = 1`, but spreading code can be skipped.
IMO comparing this nibble-split byte by non-zero values can be dangerous to miss edge cases where the upper/lower nibbles differ.

File: src/mapgen/mapgen.cpp [hunk 5]
```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
...
+				MYMAX(light_night, n.param1 & 0xF0);
+
+		n.param1 = light;
+
+		queue.emplace_back(std::make_pair(p + v3s16(0, 0, 1), light));
+		queue.emplace_back(std::make_pair(p + v3s16(0, 1, 0), light));
+		queue.emplace_back(std::make_pair(p + v3s16(1, 0, 0), light));
+		queue.emplace_back(std::make_pair(p - v3s16(0, 0, 1), light));
+		queue.emplace_back(std::make_pair(p - v3s16(0, 1, 0), light));
+		queue.emplace_back(std::make_pair(p - v3s16(1, 0, 0), light));
```
Comment (SmallJoker) [inline src/mapgen/mapgen.cpp]: Possible candidate: `g_6dirs`
https://github.com/minetest/minetest/blob/master/src/util/directiontables.cpp#L22-L31

Comment (sfan5) [reply src/mapgen/mapgen.cpp]: Perhaps, but I don't want to change or improve the algorithm itself with this PR.

**Comment by paramat** [comment]
I wonder if #4399 and #4885 are related?

> does the order in which light spreading runs matter?

In this context, i doubt it, but mostly i don't know.

Because light spread is so intensive, please can this be carefully performance-tested?

**Comment by sfan5** [comment]
> Because light spread is so intensive, please can this be carefully performance-tested?

The amount of operations performed is identical, the only performance difference can come from the "book-keeping" around it.

**Comment by paramat** [comment]
Ok, fine by me if any performance change is insignificant.

**Comment by sfan5** [comment]
Closed in favor of the other PR.

**Pull request closed without merging**